near-sdk/src/store/lookup_map/mod.rs

```diff
@@ -0,0 +1,533 @@
+        key.serialize(&mut buffer).unwrap_or_else(|_| env::panic_str(ERR_ELEMENT_SERIALIZATION));
+
+        H::hash(&buffer)
+    }
+}
+
+impl<K, V, H> LookupMap<K, V, H>
+where
+    K: BorshSerialize + Ord,
+    V: BorshSerialize + BorshDeserialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    fn deserialize_element(bytes: &[u8]) -> V {
+        V::try_from_slice(bytes).unwrap_or_else(|_| env::panic_str(ERR_ELEMENT_DESERIALIZATION))
+    }
+
+    fn load_element<Q: ?Sized>(prefix: &[u8], key: &Q) -> Option<V>
+    where
+        Q: BorshSerialize,
+        K: Borrow<Q>,
+    {
+        let storage_bytes = env::storage_read(&Self::lookup_key(prefix, key));
+        storage_bytes.as_deref().map(Self::deserialize_element)
+    }
+
+    /// Returns a reference to the value corresponding to the key.
+    ///
+    /// The key may be any borrowed form of the map's key type, but
+    /// [`BorshSerialize`] and [`ToOwned<Owned = K>`](ToOwned) on the borrowed form *must* match those for
+    /// the key type.
+    pub fn get<Q: ?Sized>(&self, k: &Q) -> Option<&V>
+    where
+        K: Borrow<Q>,
+        Q: BorshSerialize + ToOwned<Owned = K>,
+    {
+        //* ToOwned bound, which forces a clone, is required to be able to keep the key in the cache
+        let entry = self
+            .cache
+            .get(k.to_owned())
+            .get_or_init(|| CacheEntry::new_cached(Self::load_element(&self.prefix, k)));
+        entry.value().as_ref()
+    }
+
+    fn get_mut_inner<Q: ?Sized>(&mut self, k: &Q) -> &mut CacheEntry<V>
+    where
+        K: Borrow<Q>,
+        Q: BorshSerialize + ToOwned<Owned = K>,
+    {
+        let prefix = &self.prefix;
+        //* ToOwned bound, which forces a clone, is required to be able to keep the key in the cache
+        let entry = self.cache.get_mut(k.to_owned());
+        entry.get_or_init(|| CacheEntry::new_cached(Self::load_element(prefix, k)));
+        let entry = entry.get_mut().unwrap_or_else(|| unreachable!());
+        entry
+    }
+
+    /// Returns a mutable reference to the value corresponding to the key.
+    ///
+    /// The key may be any borrowed form of the map's key type, but
+    /// [`BorshSerialize`] and [`ToOwned<Owned = K>`](ToOwned) on the borrowed form *must* match those for
+    /// the key type.
+    pub fn get_mut<Q: ?Sized>(&mut self, k: &Q) -> Option<&mut V>
+    where
+        K: Borrow<Q>,
+        Q: BorshSerialize + ToOwned<Owned = K>,
+    {
+        self.get_mut_inner(k).value_mut().as_mut()
+    }
+
+    /// Inserts a key-value pair into the map.
+    ///
+    /// If the map did not have this key present, [`None`] is returned.
+    ///
+    /// If the map did have this key present, the value is updated, and the old
+    /// value is returned. The key is not updated, though; this matters for
+    /// types that can be `==` without being identical.
+    pub fn insert(&mut self, k: K, v: V) -> Option<V>
+    where
+        K: Clone,
```

Comment:
suggestion(blocking for stabilization): the `K: Clone` bound doesn't make sense -- we already have an owned `K` here. Can we write 

```rust
impl StableMap {
  fn get_key_value_mut(&mut self, k: K) -> (&mut K, &mut V) { ... }
}
```

sounds doable in theory?

Reply:
Intuitively, it makes sense, but I don't know how you're getting the `&mut K`, let alone a `&K` through the entry API like this.

I'll come back to this to see if there is anything that can be done, but at best with some code duplication all I can come up with is removing one clone, which I'm not certain is very helpful